runtime/connectors/s3/s3.go

```diff
@@ -190,10 +195,121 @@ func (c connector) HasAnonymousAccess(ctx context.Context, env *connectors.Env,
 	if err != nil {
 		return false, fmt.Errorf("failed to open bucket %q, %w", conf.url.Host, err)
 	}
+	defer bucketObj.Close()
 
 	return bucketObj.IsAccessible(ctx)
 }
 
+func (c Connector) ListBuckets(ctx context.Context, env *connectors.Env) ([]string, error) {
+	creds, err := getCredentials(env)
+	if err != nil {
+		return nil, err
+	}
+	if creds == credentials.AnonymousCredentials {
+		return nil, fmt.Errorf("no credentials exist")
+	}
+
+	// Create a session that tries to infer the region from the environment
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable, // Tells to look for default region set with `aws configure`
+		Config: aws.Config{
+			Credentials: creds,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// no region found, default to us-east-1
+	if sess.Config.Region == nil || *sess.Config.Region == "" {
+		sess = sess.Copy(&aws.Config{Region: aws.String("us-east-1")})
+	}
+	svc := s3.New(sess)
+	output, err := svc.ListBuckets(&s3.ListBucketsInput{})
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := make([]string, 0, len(output.Buckets))
+	for _, bucket := range output.Buckets {
+		if bucket.Name != nil {
+			buckets = append(buckets, *bucket.Name)
+		}
+	}
+	return buckets, nil
+}
+
+func (c Connector) ListObjects(ctx context.Context, req *runtimev1.S3ListObjectsRequest, env *connectors.Env) ([]*runtimev1.S3Object, string, error) {
+	// todo :: check for cases when accessing public buckets but env configured
+	creds, err := getCredentials(env)
+	if err != nil {
+		return nil, "", err
+	}
+
+	bucket, err := openBucket(ctx, &Config{AWSRegion: req.Region}, req.Bucket, creds)
+	if err != nil {
+		return nil, "", err
+	}
+	defer bucket.Close()
+
+	var pageToken []byte
+	if req.GetPageToken() == "" {
```

Comment:
Let's just access the properties directly instead of through getters, they're not commonly used in Go (except for in interfaces)